keps/sig-windows/20190103-windows-node-support.md

```diff
@@ -108,23 +124,33 @@ As of 29-11-2018 much of the work for enabling Windows nodes has already been co
 - Calico CNI for Pod networking
 - Hyper-V isolation (Currently this is limited to 1 container per Pod and is an alpha feature)
 - It is unclear if the RuntimeClass proposal from sig-node will simplify scheduled Windows containers. We will work with sig-node on this.
+- Properly implement terminationGracePeriodSeconds for Windows (https://github.com/moby/moby/issues/25982 and https://github.com/kubernetes/kubernetes/issues/73434)
 
 ### What will never work (Note that some features are plain unsupported while some will not work without underlying OS changes)
 - Certain Pod functionality
-    - Privileged containers and other Pod security context privilege and access control settings
+    - Privileged containers
+    - Pod security context privilege and access control settings. Any Linux Capabilities, SELinux, AppArmor, Seccomp, Capabilities (POSIX Capabilities), and others are not supported 
     - Reservations are not enforced by the OS, but overprovisioning could be blocked with `--enforce-node-allocatable=pods` (pending: tests needed)
     - Certain volume mappings
       - Single file & subpath volume mounting
+      - Subpath volume mounting for Secrets
       - Host mount projection
       - DefaultMode (due to UID/GID dependency)
       - readOnly root filesystem. Mapped volumes still support readOnly
       - Block device mapping
     - Termination Message - these require single file mappings
+    - HugePages
+    - Memory as the storage medium
 - CSI plugins, which require privileged containers
+- File system features like uui/guid, per-user Linux filesystem permissions, and read-only root filesystems
 - NFS based storage/volume support
 - Host networking is not available in Windows
+- ClusterFirstWithHostNet is not supported for DNS. Windows treats all names with a `.` as a FQDN and skips PQDN resolution
 - [Some parts of the V1 API](https://github.com/kubernetes/kubernetes/issues/70604)
+- Not all features of shared namespaces are supported. This is clarified in the API section below
+- The existing node problem detector is Linux-only and requires privileged containers. In general, we don’t expect these to be used on Windows because there’s no privileged support
 - Overlay networking support in Windows Server 1803 is not fully functional using the `win-overlay` CNI plugin. Specifically service IPs do not work on Windows nodes. This is currently specific to `win-overlay`; other CNI plugins (OVS, AzureCNI) work. Since Windows Server 1803 is not supported for GA, this is mostly not applicable. We left it here since it impacts beta
+- As noted above, there are compatibility issues enforced by Microsoft where the host OS version must match the container base image OS. Changes to this compatibility policy must come from Microsoft. For GA, since we will only support Windows Server 2019 (aka 1809), both `container host OS` and `container OS` must be running the same version of Windows, 1809.
```

Comment:
@PatrickLang @michmike Is there a plan to tackle this compatibility issue? For example, in, say, 1.17 release, will sig-windows work on `1903` and *require* `1903`, or will it support multiple OS versions (how about new features)?  Do we taint nodes based on the versions from the semi-annual channels, and keep updating taints for nodes and tolerations for pods?

/cc @dchen1107

Reply:
@yujuhong  we don't know in the future which specific operating systems we will support, but it will likely be more than one (server 2019/1809 for example as well 1903). We will need to taint nodes based on the versions of the OS. Specific PODs will be built using containers from a specific release of windows, so yes, they will need to have a specific toleration for that OS version as well

Reply:
Yes. Windows Server version 1903 would require containers built against 1903.

It is technically feasible to support a Windows Server version 1903 node running containers built with Windows Server 2019 / 1809. We should track that as a new enhancement with corresponding KEP for a future Kubernetes release. Most likely it would be implemented as part of a runtimeclass that uses Hyper-V for backwards compatibility.

Reply:
@PatrickLang @michmike 

I understand the constraint that the containers need to be built against the same OS version as the nodes (e.g., 1903). That is clear in the KEP.

It's also clear that the version supported in 1.14 will be 1809.

It sounds like @yujuhong's request is to document the expected OS version compatibility of future Kubernetes releases. For example, will 1.15 also support 1809? Will it also support 1903? What will the expected upgrade cadence be? Every Kubernetes release? Twice a year? Once a year? Is there a known, recommended upgrade procedure for upgrading user applications together with cluster nodes? Some kind of blue/green upgrade approach?

We should document what we know and what we don't know.

Reply:
And will this adhere to the project's version-skew policy that release 1.X of the cluster control plane should work with nodes of versions 1.X, 1.X-1, and 1.X-2?